core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java

```diff
@@ -337,42 +340,38 @@ void forceSorterToSpill() throws IOException {
     final int numPartitions = partitioner.numPartitions();
     final long[] partitionLengths = new long[numPartitions];
     final InputStream[] spillInputStreams = new FileInputStream[spills.length];
-    OutputStream mergedFileOutputStream = null;
+    final CountingOutputStream mergedFileOutputStream = new CountingOutputStream(
```

Comment:
Could you add a comment about why need to use `CountingOutputStream` + `CloseShieldOutputStream`? It took me a while to figure out the optimization you did.